### ISSUE TITLE
Single system dimension should not be inconsistent with the operator restrictions

### ISSUE THREAD

**Issue by cahitkargi** [CirQuS-UTS/QuanGuru#65]
If the system operator is sigmaz or some other Pauli operator, the relevant term simply calls these functions without making sure the dimension value of the system is 2.
Therefore, it creates two dimensional matrices regardless of the dimension value. It should throw an error if the operator is a Pauli but the dimension is not 2.

**Comment by cahitkargi** [comment]
This can be fixed trivially by adding an if statement in `_dimInput` function of QTerm.
More specifically, before line 272 (ie before the call to the operator `operMat = oper()**order`), it should raise an error if the dimension is not 2.

**Comment by cahitkargi** [comment]
my previous suggestion works, but would it be better to make two of such assertions ? one in dimension setter another as above. 
This is because the above case would let the dimension to be set to any value regardless of the operator, but we can further restrict that and not even set the dimension (but raise an error) if the operator is a Pauli operator

**Comment by nklangford** [comment]
I find the error message "has Pauli as term" is not very easily human readable.  It somehow relies on QuanGuru jargon, which might not be transparent to a general user.  Let's instead say something like "system operator is restricted to dimension = 2" or "system operator only works for dimension = 2" or something like that which is generically understandable by a general quantum-literate user, and doesn't require any understanding of underlying QuanGuru structure.

**Comment by AngsarM** [comment]
I don't think that Pauli is a QuanGuru jargon, and I think that it should be understandable by a general quantum-literate user. Taking your suggestion into account, what if we modify the message as: "has Pauli matrix as a term, which is restricted to dimension = 2"?

**Comment by cahitkargi** [comment]
I agree with Angsar, and I suggest something like:

```python
f"quantum system ({self.name}) has a Pauli matrix as a term, so it is restricted to dimension = 2 (cannot set the given {dim})"
```

**Comment by cahitkargi** [comment]
I think the current fix of the bug needs to be improved.

The possibilities are:
1 - Create a quantum system -> set its dimension -> set an operator 
2 - Create a quantum system -> set an operator -> set its dimension
3 - Create a quantum system -> set its dimension -> use it in a coupling
4 - Create a quantum system -> use it in a coupling -> set/change its dimension

## First case 
**The problem:**
Below code runs without any error
```python
from quanguru import QuantumSystem, sigmaz

qs = QuantumSystem()
qs.dimension = 3
qs.operator = sigmaz
```
**The current solution:**
https://github.com/CirQuS-UTS/QuanGuru/blob/73cccd2cac9769811b780bdf8c4743e095860bbc/src/quanguru/classes/QTerms.py#L272-L274
raises an error for an existing term if the dimension of the quantum-system is not 2 but the operator is one of the Pauli operators.
This lets you to create an invalid term-operator combination and raises the error only when you actually try to create the matrices (for example, with ```qs.totalHamiltonian```)

**What should happen:** 
It should already raise the error when setting the operator. I guess this means that the current solution should be moved into operator setter.

## Second case
Works as expected and immediately raises the error

## Third case
**The problem:**
Below code runs without any error
```python
from quanguru import QuantumSystem, QTerm, sigmaz

qSingle1 = QuantumSystem(dimension = 3)
qSingle2 = QuantumSystem(dimension = 3)

qComposite = QuantumSystem(subSys=[qSingle1,  qSingle2])

couplingTerm = QTerm(superSys=qComposite)
couplingTerm.qSystem = (qSingle1, qSingle1)
couplingTerm.operator = (sigmaz, sigmaz)
```
**The current solution:**
same as first case

**What should happen:** 
same as first case

## Last case
**The problem:**
Below code runs without any error
```python
from quanguru import QuantumSystem, QTerm, sigmaz

qSingle1 = QuantumSystem()
qSingle2 = QuantumSystem()

qComposite = QuantumSystem(subSys=[qSingle1, qSingle2])

couplingTerm = QTerm(superSys=qComposite, frequency=1, qSystem = (qSingle1, qSingle1), operator = (sigmaz, sigmaz))
qSingle1.dimension = 3
```

**The current solution:**
same as first case

**What should happen:** 
It should already raise an error when we try to set/change the dimension to something else than 2. When changing/setting the dimension of `self`, the dimension setter should check the whole nested-quantum-system to find any coupling that has `self` and raise an error if the corresponding coupling operator is a Pauli operator and dimension is not 2. I think this should be implemented as a separate method to be called inside dimension setter.
